lib/paperclip/storage/s3v2.rb

```diff
@@ -0,0 +1,459 @@
+        creds = find_credentials(creds).stringify_keys
+        env = Object.const_defined?(:Rails) ? Rails.env : nil
+        (creds[env] || creds).symbolize_keys
+      end
+
+      def exists?(style = default_style)
+        if original_filename
+          s3_interface.client.head_object(bucket: bucket_name, key:s3_object_name(style)).present?
+        else
+          false
+        end
+      rescue Aws::Errors::ServiceError
+        false
+      end
+
+      def s3_permissions(style = default_style)
+        s3_permissions = @s3_permissions[style] || @s3_permissions[:default]
+        s3_permissions = s3_permissions.call(self, style) if s3_permissions.respond_to?(:call)
+        s3_permissions
+      end
+
+      def s3_storage_class(style = default_style)
+        @s3_storage_class[style] || @s3_storage_class[:default]
+      end
+
+      def s3_protocol(style = default_style, with_colon = false)
+        protocol = @s3_protocol
+        protocol = protocol.call(style, self) if protocol.respond_to?(:call)
+
+        if with_colon && !protocol.empty?
+          "#{protocol}:"
+        else
+          protocol.to_s
+        end
+      end
+
+      def create_bucket
+        s3_interface.bucket(bucket_name).create
+      end
+
+      def flush_writes #:nodoc:
+        @queued_for_write.each do |style, file|
+        retries = 0
+          begin
+            log("saving #{path(style)}")
+            acl = @s3_permissions[style] || @s3_permissions[:default]
+            acl = acl.call(self, style) if acl.respond_to?(:call)
+            write_options = {
+              :content_type => file.content_type,
+              :acl => acl
+            }
+
+            # refresh headers (usefull when Proc is involved)
+            merge_s3_headers(@options[:s3_headers], @s3_headers, @s3_metadata)
+
+            # add storage class for this style if defined
+            storage_class = s3_storage_class(style)
+            write_options.merge!(:storage_class => storage_class) if storage_class
+
+            if @s3_server_side_encryption
+              write_options[:server_side_encryption] = @s3_server_side_encryption
+            end
+
+            style_specific_options = styles[style]
+
+            if style_specific_options
+              merge_s3_headers( style_specific_options[:s3_headers], @s3_headers, @s3_metadata) if style_specific_options[:s3_headers]
+              @s3_metadata.merge!(style_specific_options[:s3_metadata]) if style_specific_options[:s3_metadata]
+            end
+
+            write_options[:metadata] = @s3_metadata unless @s3_metadata.empty?
+            write_options.merge!(@s3_headers)
+
+            s3_object(style).upload_file(file.path, write_options)
+          rescue Aws::S3::Errors::NoSuchBucket
+            create_bucket
+            retry
+          rescue Aws::S3::Errors::SlowDown    # Daniel: SlowDown not defined in V2. See what concept replaces it. There is a Waiters concept
```

Comment:
Actually the Watchers are a new concept but don't seem to match the us of the SlowDown exception and I din't think we can use them to the same end. My thought was to just remove the SlowDown exception rescuing as it's not thrown by the V2 api anymore.  Note the V2 SDK has now a built in retry_limit (3 by default).